diff --git a/config.yml.example b/config.yml.example
index 97b5006..000ba4b 100644
--- a/config.yml.example
+++ b/config.yml.example
@@ -25,6 +25,10 @@ repos_path: "/home/git/repositories"
 # File used as authorized_keys for gitlab user
 auth_file: "/home/git/.ssh/authorized_keys"
 
+# File that contains the secret key for verifying access to GitLab.
+# Default is .gitlab_shell_secret in the root directory.
+# secret_file: "/home/git/gitlab-shell/.gitlab_shell_secret"
+
 # Redis settings used for pushing commit notices to gitlab
 redis:
   bin: /usr/bin/redis-cli
diff --git a/lib/gitlab_config.rb b/lib/gitlab_config.rb
index c97743b..1d8c690 100644
--- a/lib/gitlab_config.rb
+++ b/lib/gitlab_config.rb
@@ -19,6 +19,10 @@ def auth_file
     @config['auth_file'] ||= File.join(home, ".ssh/authorized_keys")
   end
 
+  def secret_file
+    @config['secret_file'] ||= File.join(ROOT_PATH, '.gitlab_shell_secret')
+  end
+
   def gitlab_url
     @config['gitlab_url'] ||= "http://localhost/"
   end
diff --git a/lib/gitlab_net.rb b/lib/gitlab_net.rb
index c178927..bc9339b 100644
--- a/lib/gitlab_net.rb
+++ b/lib/gitlab_net.rb
@@ -125,6 +125,6 @@ def cert_store
   end
 
   def secret_token
-    @secret_token ||= File.read File.join(ROOT_PATH, '.gitlab_shell_secret')
+    @secret_token ||= File.read config.secret_file
   end
 end
