--- a/portal-impl/src/com/liferay/portal/tools/deploy/BaseDeployer.java
+++ b/portal-impl/src/com/liferay/portal/tools/deploy/BaseDeployer.java
@@ -34,6 +34,8 @@
 import com.liferay.portal.kernel.util.CharPool;
 import com.liferay.portal.kernel.util.FileUtil;
 import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.Html;
+import com.liferay.portal.util.HtmlImpl;
 import com.liferay.portal.kernel.util.HttpUtil;
 import com.liferay.portal.kernel.util.OSDetector;
 import com.liferay.portal.kernel.util.PropertiesUtil;
@@ -2139,6 +2140,10 @@
 			content = content.substring(0, x) + content.substring(y);
 		}
 
+		// strip comments, they are on the end stripped anyway
+		Html html = new HtmlImpl();
+		content = html.stripComments(content);
+
 		Document document = SAXReaderUtil.read(content);
 
 		Element rootElement = document.getRootElement();
@@ -2206,10 +2211,12 @@
 		String extraContent = getExtraContent(
 			webXmlVersion, srcFile, displayName);
 
-		int pos = content.indexOf("<listener>");
+		int pos = content.lastIndexOf("</listener>");
 
 		if (pos == -1) {
 			pos = content.indexOf("</web-app>");
+		} else {
+			pos += 11;
 		}
 
 		String newContent =
